Use `download-rustc=false` global default, `if-unchanged` for tools and library profiles, and make `rust.debug-assertions=true` inhibit downloading CI rustc

jieyouxu opened the pull request:
- Use `download-rustc = false` as global default.
    - Use `download-rustc = 'if-unchanged'` for tools and library profiles.
- Make `rust.debug-assertions = true` inhibit downloading CI rustc because alt rustc builds do not yet have rustc debug assertions enabled.

Fixes #133132.

cc discussions: https://rust-lang.zulipchat.com/#narrow/channel/326414-t-infra.2Fbootstrap/topic/Bootstrap.20breakage
compiler contributors poll: https://rust-lang.zulipchat.com/#narrow/channel/131828-t-compiler/topic/.60download-rustc.20.3D.20'if-unchanged'.60.20for.20.60compiler.60.20profile.3F/near/481877253
library contributors poll: https://rust-lang.zulipchat.com/#narrow/channel/219381-t-libs/topic/.60download-rustc.20.3D.20.22if-unchanged.22.60.20default.20for.20libs.20profile.3F/near/482607011
cc https://rust-lang.zulipchat.com/#narrow/channel/326414-t-infra.2Fbootstrap/topic/When.20is.20rustc.20built.20with.20debug.20assertions.3F

cc @MarcoIeni since you're working on improving CI job times, sorry, this will definitely regress some CI job times because we're probably lying to ourselves that CI rustc had debug assertions for some time 😅

cc @onur-ozkan for FYI, but since you're on vacation (sorry for the ping),

r? @Kobzol (I *think* you have a bit more context than other bootstrap reviewers?)

jieyouxu commented:
I think the motivation for `download-rustc = 'if-unchanged'` for library profile is to prepare for the staging redesign, not 100% certain.

jieyouxu commented:
Waiting for libs contributors poll to determine if we want to set the library profile to use `if-unchanged`.
@rustbot blocked

jieyouxu commented:
This seems mostly uncontroversial to use `if-unchanged` for library profile after improved detection logic in #131831.

@rustbot ready

jieyouxu committed:
bootstrap: make `rust.debug-assertions = true` inhibit downloading CI rustc

Warn if `rust.download-rustc = true` is used with
`rust.debug-assertions` as alt CI rustc is not currently built with
debug assertions (not to be confused with LLVM assertions).

jieyouxu commented:
Maybe r? @clubby789 (Onur is on vacation and Jakub/Mark/Albert is busy AFAIK)

jieyouxu commented on src/bootstrap/src/core/config/config.rs:
```diff
@@ -1767,6 +1767,29 @@ impl Config {
+            //
+            // This relies also on the fact that global default for `download-rustc` will be `false`
+            // if it's not explicitly set.
+            if matches!(rustc_debug_assertions, Some(true))
+                || (matches!(debug, Some(true)) && rustc_debug_assertions.is_none())
+            {
```
I don't like this but this is the simplest thing I could think of without fixing our config layering

clubby789 commented on src/bootstrap/src/core/config/config.rs:
```diff
@@ -2782,8 +2805,12 @@ impl Config {
-            Some(StringOrBool::Bool(false)) => return None,
+            // Globally default for `download-rustc` to `false`, because some contributors don't use
+            // profiles for reasons such as:
+            // - They need to seemlessly switch between compiler/library work.
```
nit: `seamlessly`

jieyouxu replied:
Fixed.

clubby789 reviewed (approved):
Seems reasonable, r=me with typo fix

EDIT: There's also a leftover 'for' a couple of lines above

jieyouxu commented:
Only change is typo fix ([diff](https://github.com/rust-lang/rust/compare/9f1063572c5de5a35a5d8c43da1a23a3484b99c5..1c4940e7bec357f0bd0bdad77d5513b65b5ee6ed)).

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11880042361/job/33102706320) [(plain)](https://github.com/rust-lang/rust/commit/1c4940e7bec357f0bd0bdad77d5513b65b5ee6ed/checks/33102706320/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 exporting to docker image format
#16 sending tarball 27.5s done
#16 DONE 40.9s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.93s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
...........................

failures:

---- core::config::tests::download_rustc_xor_rustc_debug_assertions_both_true stdout ----
warning: `download-rustc` is enabled, but there are changes to one of [":!src/tools", ":!tests", ":!triagebot.toml"]
NOTE: test did not panic as expected
---- core::config::tests::download_rustc_xor_rustc_debug_assertions_debug_download_unchanged stdout ----
NOTE: test did not panic as expected
---- core::config::tests::download_rustc_xor_rustc_debug_assertions_debug_true stdout ----
---- core::config::tests::download_rustc_xor_rustc_debug_assertions_debug_true stdout ----
warning: `download-rustc` is enabled, but there are changes to one of [":!src/tools", ":!tests", ":!triagebot.toml"]
NOTE: test did not panic as expected

failures:
    core::config::tests::download_rustc_xor_rustc_debug_assertions_both_true
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11880860270/job/33104516189) [(plain)](https://github.com/rust-lang/rust/commit/53cbe938b87762795017efe16c18d6507fdb9bdb/checks/33104516189/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling build_helper v0.1.0 (/checkout/src/build_helper)
    Finished `dev` profile [unoptimized] target(s) in 14.94s
##[endgroup]
thread 'main' panicked at src/core/config/config.rs:1785:25:
ERROR: currently no CI rustc builds have rustc debug assertionsenabled. Please either set `rust.debug-assertions` to `false` or set `rust.download-rustc` to `false`.
Build completed unsuccessfully in 0:00:25
make: *** [Makefile:98: prepare] Error 1
Command failed. Attempt 2/5:
##[group]Building bootstrap
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
thread 'main' panicked at src/core/config/config.rs:1785:25:
ERROR: currently no CI rustc builds have rustc debug assertionsenabled. Please either set `rust.debug-assertions` to `false` or set `rust.download-rustc` to `false`.
Build completed unsuccessfully in 0:00:00
make: *** [Makefile:98: prepare] Error 1
Command failed. Attempt 3/5:
##[group]Building bootstrap
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
thread 'main' panicked at src/core/config/config.rs:1785:25:
ERROR: currently no CI rustc builds have rustc debug assertionsenabled. Please either set `rust.debug-assertions` to `false` or set `rust.download-rustc` to `false`.
Build completed unsuccessfully in 0:00:00
make: *** [Makefile:98: prepare] Error 1
Command failed. Attempt 4/5:
##[group]Building bootstrap
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
thread 'main' panicked at src/core/config/config.rs:1785:25:
ERROR: currently no CI rustc builds have rustc debug assertionsenabled. Please either set `rust.debug-assertions` to `false` or set `rust.download-rustc` to `false`.
Build completed unsuccessfully in 0:00:00
make: *** [Makefile:98: prepare] Error 1
Command failed. Attempt 5/5:
##[group]Building bootstrap
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
thread 'main' panicked at src/core/config/config.rs:1785:25:
ERROR: currently no CI rustc builds have rustc debug assertionsenabled. Please either set `rust.debug-assertions` to `false` or set `rust.download-rustc` to `false`.
Build completed unsuccessfully in 0:00:00
make: *** [Makefile:98: prepare] Error 1
The command has failed after 5 attempts.
  local time: Sun Nov 17 17:36:11 UTC 2024
```

</details>

jieyouxu committed:
ci: don't force CI rustc for `x86_64-gnu-tools`

CI alt rustc builds do not have rustc debug assertions currently.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11881392120/job/33105678518) [(plain)](https://github.com/rust-lang/rust/commit/4fd0d92160d7077a2fa91d4a8359ea83295323d5/checks/33105678518/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling termcolor v1.4.1
    Finished `dev` profile [unoptimized] target(s) in 24.97s
##[endgroup]
thread 'main' panicked at src/core/config/config.rs:1785:25:
ERROR: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` or set `rust.download-rustc` to `false`.
Build completed unsuccessfully in 0:00:35
make: *** [Makefile:98: prepare] Error 1
Command failed. Attempt 2/5:
##[group]Building bootstrap
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
thread 'main' panicked at src/core/config/config.rs:1785:25:
ERROR: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` or set `rust.download-rustc` to `false`.
Build completed unsuccessfully in 0:00:00
make: *** [Makefile:98: prepare] Error 1
Command failed. Attempt 3/5:
##[group]Building bootstrap
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
thread 'main' panicked at src/core/config/config.rs:1785:25:
ERROR: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` or set `rust.download-rustc` to `false`.
Build completed unsuccessfully in 0:00:00
make: *** [Makefile:98: prepare] Error 1
Command failed. Attempt 4/5:
##[group]Building bootstrap
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
thread 'main' panicked at src/core/config/config.rs:1785:25:
ERROR: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` or set `rust.download-rustc` to `false`.
Build completed unsuccessfully in 0:00:00
make: *** [Makefile:98: prepare] Error 1
Command failed. Attempt 5/5:
##[group]Building bootstrap
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
thread 'main' panicked at src/core/config/config.rs:1785:25:
ERROR: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` or set `rust.download-rustc` to `false`.
Build completed unsuccessfully in 0:00:00
make: *** [Makefile:98: prepare] Error 1
The command has failed after 5 attempts.
  local time: Sun Nov 17 19:01:45 UTC 2024
```

</details>

jieyouxu commented:
Yeah, like I suspected, this is going to be problematic.

@rustbot blocked (https://rust-lang.zulipchat.com/#narrow/channel/242791-t-infra/topic/CI.20jobs.20with.20.60--enable-debug-assertions.60.20.2B.20CI.20rustc)

jieyouxu commented:
This will also need someone from infra to double-check... so r? @Mark-Simulacrum (or @Kobzol or infra)

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11891888137/job/33133588945) [(plain)](https://github.com/rust-lang/rust/commit/c1e065ace2c731fb1fd4d54142ebeb9f1c40633f/checks/33133588945/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling home v0.5.9
    Finished `dev` profile [unoptimized] target(s) in 25.74s
##[endgroup]
thread 'main' panicked at src/core/config/config.rs:1785:25:
ERROR: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` or set `rust.download-rustc` to `false`.
Build completed unsuccessfully in 0:00:36
make: *** [Makefile:98: prepare] Error 1
Command failed. Attempt 2/5:
##[group]Building bootstrap
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
thread 'main' panicked at src/core/config/config.rs:1785:25:
ERROR: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` or set `rust.download-rustc` to `false`.
Build completed unsuccessfully in 0:00:00
make: *** [Makefile:98: prepare] Error 1
Command failed. Attempt 3/5:
##[group]Building bootstrap
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
thread 'main' panicked at src/core/config/config.rs:1785:25:
ERROR: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` or set `rust.download-rustc` to `false`.
Build completed unsuccessfully in 0:00:00
make: *** [Makefile:98: prepare] Error 1
Command failed. Attempt 4/5:
##[group]Building bootstrap
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
thread 'main' panicked at src/core/config/config.rs:1785:25:
ERROR: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` or set `rust.download-rustc` to `false`.
Build completed unsuccessfully in 0:00:00
make: *** [Makefile:98: prepare] Error 1
Command failed. Attempt 5/5:
##[group]Building bootstrap
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
thread 'main' panicked at src/core/config/config.rs:1785:25:
ERROR: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` or set `rust.download-rustc` to `false`.
Build completed unsuccessfully in 0:00:00
make: *** [Makefile:98: prepare] Error 1
The command has failed after 5 attempts.
  local time: Mon Nov 18 12:02:31 UTC 2024
```

</details>

jieyouxu commented:
For `mingw-check`, AFAICT this is intended for use as a fast-rejection job, so for this specifically I set `NO_DEBUG_ASSERTIONS=1`, while for `x86_64-gnu-tools` I removed `FORCE_CI_RUSTC`.

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11892172068/job/33134476620) [(plain)](https://github.com/rust-lang/rust/commit/ef5d82e941db5fb06d6546493c705e1ed9411ce9/checks/33134476620/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.823 Building wheels for collected packages: reuse
#13 2.824   Building wheel for reuse (pyproject.toml): started
#13 3.071   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.072   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132720 sha256=026f3bb0f1aa8090b861fd0a0939cb1a782396d84c8aab7875096557d637a0f6
#13 3.073   Stored in directory: /tmp/pip-ephem-wheel-cache-yv48cju6/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.075 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.484 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.484 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 4.016 Collecting virtualenv
#13 4.016 Collecting virtualenv
#13 4.073   Downloading virtualenv-20.27.1-py3-none-any.whl (3.1 MB)
#13 4.278      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.1/3.1 MB 15.3 MB/s eta 0:00:00
#13 4.314 Collecting distlib<1,>=0.3.7
#13 4.324   Downloading distlib-0.3.9-py2.py3-none-any.whl (468 kB)
#13 4.342      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 27.4 MB/s eta 0:00:00
#13 4.374 Collecting platformdirs<5,>=3.9.1
#13 4.384   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#13 4.421 Collecting filelock<4,>=3.12.2
#13 4.431   Downloading filelock-3.16.1-py3-none-any.whl (16 kB)
#13 4.511 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.690 Successfully installed distlib-0.3.9 filelock-3.16.1 platformdirs-4.3.6 virtualenv-20.27.1
#13 DONE 4.8s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
   Compiling build_helper v0.1.0 (/checkout/src/build_helper)
    Finished `dev` profile [unoptimized] target(s) in 25.48s
##[endgroup]
thread 'main' panicked at src/core/config/config.rs:1785:25:
ERROR: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` or set `rust.download-rustc` to `false`.
Build completed unsuccessfully in 0:00:36
make: *** [Makefile:98: prepare] Error 1
Command failed. Attempt 2/5:
##[group]Building bootstrap
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
thread 'main' panicked at src/core/config/config.rs:1785:25:
ERROR: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` or set `rust.download-rustc` to `false`.
Build completed unsuccessfully in 0:00:00
make: *** [Makefile:98: prepare] Error 1
Command failed. Attempt 3/5:
##[group]Building bootstrap
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
thread 'main' panicked at src/core/config/config.rs:1785:25:
ERROR: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` or set `rust.download-rustc` to `false`.
Build completed unsuccessfully in 0:00:00
make: *** [Makefile:98: prepare] Error 1
Command failed. Attempt 4/5:
##[group]Building bootstrap
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
thread 'main' panicked at src/core/config/config.rs:1785:25:
ERROR: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` or set `rust.download-rustc` to `false`.
Build completed unsuccessfully in 0:00:00
make: *** [Makefile:98: prepare] Error 1
Command failed. Attempt 5/5:
##[group]Building bootstrap
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
thread 'main' panicked at src/core/config/config.rs:1785:25:
ERROR: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` or set `rust.download-rustc` to `false`.
Build completed unsuccessfully in 0:00:00
make: *** [Makefile:98: prepare] Error 1
The command has failed after 5 attempts.
  local time: Mon Nov 18 12:18:52 UTC 2024
```

</details>

jieyouxu commented:
Hm this approach is problematic. I need to change the logic such that `rust.debug-assertions=true` inhibits `download-rustc` and force a fresh build.

bors commented:
:umbrella: The latest upstream changes (presumably #133219) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

MarcoIeni commented:
> cc @MarcoIeni since you're working on improving CI job times, sorry, this will definitely regress some CI job times 

Thanks for letting me know! I will have a look at the job timings after this one is merged to make sure the auto build has reasonable timing 👍

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11952170276/job/33317347656) [(plain)](https://github.com/rust-lang/rust/commit/de92e3e687693e631d49e160e466fbf1bcfd734f/checks/33317347656/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 exporting to docker image format
#16 sending tarball 28.8s done
#16 DONE 35.1s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.99s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
...........................

failures:

---- core::config::tests::download_rustc_xor_rustc_debug_assertions_both_true stdout ----
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` or set `rust.download-rustc` to `false` if you want to use download CI rustc.
warning: `download-rustc` is enabled, but there are changes to one of [":!src/tools", ":!tests", ":!triagebot.toml"]
WARN: `rust.debug-assertions = true` will prevent downloading CI rustc as alt CI rustc is not currently built with debug assertions.
NOTE: test did not panic as expected
---- core::config::tests::download_rustc_xor_rustc_debug_assertions_debug_download_unchanged stdout ----
---- core::config::tests::download_rustc_xor_rustc_debug_assertions_debug_download_unchanged stdout ----
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` or set `rust.download-rustc` to `false` if you want to use download CI rustc.
NOTE: test did not panic as expected
---- core::config::tests::download_rustc_xor_rustc_debug_assertions_debug_true stdout ----
---- core::config::tests::download_rustc_xor_rustc_debug_assertions_debug_true stdout ----
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` or set `rust.download-rustc` to `false` if you want to use download CI rustc.
warning: `download-rustc` is enabled, but there are changes to one of [":!src/tools", ":!tests", ":!triagebot.toml"]
WARN: `rust.debug-assertions = true` will prevent downloading CI rustc as alt CI rustc is not currently built with debug assertions.
NOTE: test did not panic as expected

failures:
    core::config::tests::download_rustc_xor_rustc_debug_assertions_both_true
```

</details>

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11953127458/job/33320363032) [(plain)](https://github.com/rust-lang/rust/commit/4999c0fc2d45e59cc3db580b49da07b4555d05bc/checks/33320363032/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.740 Building wheels for collected packages: reuse
#13 2.741   Building wheel for reuse (pyproject.toml): started
#13 2.988   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 2.989   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132720 sha256=026f3bb0f1aa8090b861fd0a0939cb1a782396d84c8aab7875096557d637a0f6
#13 2.990   Stored in directory: /tmp/pip-ephem-wheel-cache-euoqjh2x/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 2.992 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.378 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.378 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 3.912 Collecting virtualenv
#13 3.912 Collecting virtualenv
#13 4.037   Downloading virtualenv-20.27.1-py3-none-any.whl (3.1 MB)
#13 4.235      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.1/3.1 MB 16.1 MB/s eta 0:00:00
#13 4.291 Collecting filelock<4,>=3.12.2
#13 4.322   Downloading filelock-3.16.1-py3-none-any.whl (16 kB)
#13 4.356 Collecting platformdirs<5,>=3.9.1
#13 4.387   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#13 4.405 Collecting distlib<1,>=0.3.7
#13 4.444      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 88.7 MB/s eta 0:00:00
#13 4.444      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 88.7 MB/s eta 0:00:00
#13 4.524 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.702 Successfully installed distlib-0.3.9 filelock-3.16.1 platformdirs-4.3.6 virtualenv-20.27.1
#13 DONE 4.8s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
   Compiling build_helper v0.1.0 (/checkout/src/build_helper)
    Finished `dev` profile [unoptimized] target(s) in 26.05s
##[endgroup]
thread 'main' panicked at src/core/config/config.rs:1790:25:
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
Build completed unsuccessfully in 0:00:36
make: *** [Makefile:98: prepare] Error 1
Command failed. Attempt 2/5:
##[group]Building bootstrap
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
thread 'main' panicked at src/core/config/config.rs:1790:25:
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
Build completed unsuccessfully in 0:00:00
make: *** [Makefile:98: prepare] Error 1
Command failed. Attempt 3/5:
##[group]Building bootstrap
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
thread 'main' panicked at src/core/config/config.rs:1790:25:
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
Build completed unsuccessfully in 0:00:00
make: *** [Makefile:98: prepare] Error 1
Command failed. Attempt 4/5:
##[group]Building bootstrap
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
thread 'main' panicked at src/core/config/config.rs:1790:25:
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
Build completed unsuccessfully in 0:00:00
make: *** [Makefile:98: prepare] Error 1
Command failed. Attempt 5/5:
##[group]Building bootstrap
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
thread 'main' panicked at src/core/config/config.rs:1790:25:
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
Build completed unsuccessfully in 0:00:00
make: *** [Makefile:98: prepare] Error 1
The command has failed after 5 attempts.
  local time: Thu Nov 21 12:30:58 UTC 2024
```

</details>

clubby789 commented on src/bootstrap/src/core/config/config.rs:
```diff
@@ -1771,8 +1771,36 @@ impl Config {
+            if debug_assertions_requested {
+                if let Some(ref opt) = download_rustc {
+                    if opt.is_string_or_true() {
+                        println!(
```
nit: We're not very consistent about this within bootstrap, but IMO messages for user consumption should be `eprintln`'d so they don't get put into `grep` etc. This is mostly a preference thing and it's currently pretty 50/50 so up to you if you want to change this

jieyouxu replied:
Fixed.

clubby789 commented on src/bootstrap/src/core/config/config.rs:
```diff
@@ -2845,6 +2878,14 @@ impl Config {
         }
 
+        if debug_assertions_requested {
+            println!(
```
nit: same as above `eprintln` note

jieyouxu replied:
Fixed.

clubby789 commented:
LGTM, r=me with or without comments addressed

jieyouxu commented:
I'll adjust this PR tmrw, thanks for the review.

jieyouxu committed:
bootstrap: revert `rust.download-rustc` global default to `false`

And only default library and tools profile to `rust.download-rustc =
"if-unchanged"`.

jieyouxu committed:
bootstrap: add change tracker entry

jieyouxu commented on src/bootstrap/src/core/config/config.rs:
```diff
@@ -1771,8 +1771,36 @@ impl Config {
+            // Before something like [Enable debug assertions on alt
+            // builds](https://github.com/rust-lang/rust/pull/131077) lands, alt rustc builds do
+            // *not* have rustc debug assertions enabled. We should not download an CI alt rustc if
+            // we need rustc to have debug assertions (e.g. for crashes test suite).
```
Modified to use this better wording and opened #133381 to track this.

jieyouxu commented:
### Changes since last review

- Swapped to `eprintln!` (both instances).
- Reworded comment and linked FIXME to new issue #133381.
- Pulled out change tracker entry into its own commit, and fixed the wording to more accurately reflect current logic.

jieyouxu commented:
Review comments addressed, this has no other significant changes since last review, so
@bors r=@clubby789 rollup=never (significant bootstrap/CI changes)

bors commented:
:pushpin: Commit f7f8ad92a1a109229264f1a160d4b8cac3fe395b has been approved by `clubby789`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=clubby789 f7f8ad92a1a109229264f1a160d4b8cac3fe395b -->
<!-- homu: {"type":"Approved","sha":"f7f8ad92a1a109229264f1a160d4b8cac3fe395b","approver":"clubby789","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

jieyouxu commented:
Actually I should wait for PR CI.
@bors r-

jieyouxu commented:
@bors r=@clubby789 rollup=never

bors commented:
:pushpin: Commit f7f8ad92a1a109229264f1a160d4b8cac3fe395b has been approved by `clubby789`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=clubby789 f7f8ad92a1a109229264f1a160d4b8cac3fe395b -->
<!-- homu: {"type":"Approved","sha":"f7f8ad92a1a109229264f1a160d4b8cac3fe395b","approver":"clubby789","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit f7f8ad92a1a109229264f1a160d4b8cac3fe395b with merge c16c0f62ebdf72b0ec0dc55080635c02969ab6b9...
<!-- homu: {"type":"BuildStarted","head_sha":"f7f8ad92a1a109229264f1a160d4b8cac3fe395b","merge_sha":"c16c0f62ebdf72b0ec0dc55080635c02969ab6b9"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-aux`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/11996478788/job/33440974500) [(plain)](https://github.com/rust-lang-ci/rust/commit/c16c0f62ebdf72b0ec0dc55080635c02969ab6b9/checks/33440974500/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/11996478788/job/33442598697)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/11996478788/job/33442598697","builder_name":"checks-actions"} -->

jieyouxu commented:
@bors retry (spurious miri failure)

bors commented:
:hourglass: Testing commit f7f8ad92a1a109229264f1a160d4b8cac3fe395b with merge 481b5fadd7994d0f04e9a6fe9ded3f22d6753825...
<!-- homu: {"type":"BuildStarted","head_sha":"f7f8ad92a1a109229264f1a160d4b8cac3fe395b","merge_sha":"481b5fadd7994d0f04e9a6fe9ded3f22d6753825"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/11997778000/job/33446050291)
Approved by: clubby789
Pushing 481b5fadd7994d0f04e9a6fe9ded3f22d6753825 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"clubby789","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/11997778000/job/33446050291"},"merge_sha":"481b5fadd7994d0f04e9a6fe9ded3f22d6753825"} -->
